### PR TITLE
Redis(Cache) now (re)supports function as a key_prefix

### LINKED ISSUE #331: Redis doesn't supports function as a key_prefix anymore
The functionality was added here : https://github.com/pallets-eco/flask-caching/pull/109 but removed when using cachelib as backend https://github.com/pallets-eco/flask-caching/pull/308/files#diff-d15a59e5c66734a6b6d638efe7582b3a147d5a4213dc3b3ca4e3672c50607ae1.

### LINKED ISSUE #529: RedisCache support for callable key prefixes was removed as of 1.11.0
In version 1.10.1, `RedisCache` supported a callable key prefix being passed, thanks to the following `_get_prefix` function:
```python
def _get_prefix(self):
    return (
        self.key_prefix if isinstance(self.key_prefix, str) else self.key_prefix()
    )
```
However in 1.11.0, https://github.com/pallets-eco/flask-caching/pull/308 replaced usages of this function with implementations from `cachelib`, which does not have such a function for supporting a callable key prefix.

Environment:

- Python version: N/A
- Flask-Caching version: >1.11.0

### LINKED ISSUE #536: key_prefix can support func ？ : Expected type 'str', got '() -> str' instead #208
```python
def key_str():
    key = flask.request.path
    return key

@app.route('/get', methods=['get'])
@cache.cached(timeout=300, key_prefix=key_str)

```
image

when I use
```python
@cache.cached(timeout=300, key_prefix=key_str())
```

image

### LINKED ISSUE #27077: RedisCache lost support for function as a key_prefix
### Bug description

The hability to specify a function as a key_prefix for RedisCache doesn't work anymore in Superset.
`Expected type 'str', got '() -> str' instead` error is thrown.

This was fundamental for my usecase where I use the subdomain of the Superset URL as an ingredient of my PostgreSQL RLS.
Each subdomain correspond to a subset of the data that users may query through Superset (only one instance).

Introducing the notion of subdomain in the Redis cache key prefix avoids a cache mismatch between subdomains and data subsets.
It's implemented with a function parametrized in RedisCache key_prefix.

I submitted a PR in the concerned lib project : pallets-eco/cachelib#332.

### How to reproduce the bug

1. Create a simple function returning a string.
2. Parameter this function in the RedisCache key_prefix.
3. Display in Superset anything related to your data that will be cached to throw the error.

### Screenshots/recordings

_No response_

### Superset version

3.0.3

### Python version

3.11

### Node version

Not applicable

### Browser

Not applicable

### Additional context

`Expected type 'str', got '() -> str'

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

### PR DESCRIPTION
**Pull request by dbascoules**
- Behaviour added here : https://github.com/pallets-eco/flask-caching/pull/109
 - Behaviour removed here : https://github.com/pallets-eco/flask-caching/pull/308/files#diff-d15a59e5c66734a6b6d638efe7582b3a147d5a4213dc3b3ca4e3672c50607ae1

- fixes #331
- fixes pallets-eco/flask-caching#536
- fixes pallets-eco/flask-caching#529
- fixes apache/superset#27077

### CODE REVIEW

**Comment by northernSage** [comment]
Hey @dbascoules, thanks for taking the time to open the PR. Changes seem good overall, we are just a short of a test case covering the new feature. Would you mind adding one? (just let me know if you can't I'll pick it up, no worries).

File: src/cachelib/redis.py [hunk 1]
```diff
@@ -57,6 +57,11 @@ def __init__(
             self._read_client = self._write_client = host
         self.key_prefix = key_prefix or ""
 
+    def _get_prefix(self):
+        return (
+            self.key_prefix if isinstance(self.key_prefix, str) else self.key_prefix()
```
Comment (northernSage) [inline src/cachelib/redis.py:62]: Not really an issue here, but I wonder if using On a side note, I wonder if using https://docs.python.org/3/library/functions.html#callable wouldn't be make more sense semantically. In the end we want to call the key if it's callable and use it as is if it's not

**Review by northernSage — CHANGES_REQUESTED** [review]

**Comment by northernSage** [comment]
close in favour of #332 with a more descriptive branch naming

**Comment by dbascoules** [comment]
Thank you for your review and maintaining this library.

**Pull request closed without merging**